jsonrpc/eth_state_test.go

```diff
@@ -580,94 +581,196 @@ func TestEth_State_GetStorageAt(t *testing.T) {
+		},
+		{
+			"insufficient gas limit from the transaction",
+			state.TxGas,
+			ErrGasCapOverflow,
+			constructMockTx(argUintPtr(state.TxGas/2), nil),
+		},
+	}
+
+	for _, testCase := range testTable {
+		t.Run(testCase.name, func(t *testing.T) {
+			// Set up the apply hook
+			if errors.Is(testCase.expectedError, ErrGasCapOverflow) {
+				// We want to trigger a situation where no value in the gas range is correct
+				store.applyTxnHook = func(
+					header *types.Header,
+					txn *types.Transaction,
+				) (*runtime.ExecutionResult, error) {
+					return &runtime.ExecutionResult{}, state.ErrNotEnoughIntrinsicGas
+				}
+			} else {
+				// We want to trigger a situation where only values that cover the intrinsic gas costs
+				// are correct
+				store.applyTxnHook = func(
+					header *types.Header,
+					txn *types.Transaction,
+				) (*runtime.ExecutionResult, error) {
+					if txn.Gas < testCase.intrinsicGasCost {
+						return &runtime.ExecutionResult{}, state.ErrNotEnoughIntrinsicGas
+					}
+
+					return &runtime.ExecutionResult{}, nil
+				}
+			}
+
+			// Run the estimation
+			estimate, estimateErr := ethEndpoint.EstimateGas(testCase.transaction, nil)
+
+			if testCase.expectedError != nil {
+				if estimateErr == nil {
+					t.Fatalf("no error occurred, although it was expected")
+				}
+
+				// Make sure the expected errors appear
+				assert.ErrorAs(t, estimateErr, &testCase.expectedError)
+
+				// Make sure the estimate is nullified
+				assert.Equal(t, 0, estimate)
+			} else {
+				// Make sure no errors occurred
+				assert.NoError(t, estimateErr)
+
+				// Make sure the estimate is correct
+				assert.Equal(t, fmt.Sprintf("0x%x", testCase.intrinsicGasCost), estimate)
+			}
+		})
+	}
+}
+
+func TestEth_EstimateGas_Reverts(t *testing.T) {
+	// Example revert data that has the string "revert reason" as the revert reason
+	exampleReturnData := "08c379a000000000000000000000000000000000000000000000000000000000000000" +
+		"20000000000000000000000000000000000000000000000000000000000000000d72657665727420726561736f6e" +
+		"00000000000000000000000000000000000000"
+	rawReturnData, err := hex.DecodeHex(exampleReturnData)
+	assert.NoError(t, err)
+
+	revertReason := errors.New("revert reason")
+
+	store := getExampleStore()
+	ethEndpoint := newTestEthEndpoint(store)
+
+	// We want to simulate an EVM revert here
+	store.applyTxnHook = func(
+		header *types.Header,
+		txn *types.Transaction,
+	) (*runtime.ExecutionResult, error) {
+		return &runtime.ExecutionResult{
+			ReturnValue: rawReturnData,
+			Err:         runtime.ErrExecutionReverted,
+		}, nil
+	}
+
+	// Run the estimation
+	estimate, estimateErr := ethEndpoint.EstimateGas(
+		constructMockTx(nil, nil),
+		nil,
+	)
+
+	assert.Equal(t, 0, estimate)
+
+	// Make sure the EVM revert message is contained
+	assert.ErrorAs(t, estimateErr, &runtime.ErrExecutionReverted)
+
+	// Make sure the EVM revert reason is contained
+	assert.ErrorAs(t, estimateErr, &revertReason)
```

Comment:
You can use `ErrorIs` instead of `ErrorAs` except for this line (because revertReason is created ad-hoc)
Why you don't want to use `ErrorIs`?

Reply:
Answered in https://github.com/0xPolygon/polygon-edge/pull/422#discussion_r818489040